### PR TITLE
[New Term] Java ArrayList: `.indexOf()`

### LINKED ISSUE #1453: [Term Entry] Java ArrayList: `.indexOf()`
### Reason/inspiration (optional)

Add a new term entry under the [Java: ArrayList](https://www.codecademy.com/resources/docs/java/array-list) concept. This entry should include:

- An explanation of the method.
- A "Syntax" section with a ```pseudo code snippet showing the general syntax.
- An "Example" section with a ```java code snippet showing how it works.

### Entry Type (select all that apply)

New Term for Existing Concept Entry

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [x] Issue is assigned

### PR DESCRIPTION
**Pull request by Pavan-Nambi**
### Description

Added new term `.indexOf()`  method of ArrayList

Closes #1453 

### Type of Change

- Adding a new entry
- Updating the documentation

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

### CODE REVIEW

**Comment by Dusch4593** [comment]
Thanks for submitting this, @Pavan-Nambi. I'll take a look at this for review 1️⃣ 😄

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:15]:
```suggestion
The **`.indexOf()`** method returns the index of the first occurrence of the specified element in an `ArrayList`. If no element is found, -1 is returned.
```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:9]:
```suggestion
  - 'Data Types'
  - 'Elements'
  - 'Search'
```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:3]:
```suggestion
Description: 'Returns the index of the first occurrence of a given element, or -1 if not found.'
```

Comment (Dusch4593) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:3]: Since this is a term entry, the `Description` should begin with a present-tense verb.

**Comment by Pavan-Nambi** [comment]
hey ok thankyou and is there any plans on adding rust? @Dusch4593

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:15]: Since this term is already under `ArrayList`, there is no need to link to the parent concept entry. 😄

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:20]:
```suggestion
myArrayList.indexOf(element);
```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:27]:
```suggestion
If the `element` exists in `myArrayList`, the `.indexOf()` method will return the index of its first occurrence (even if the value if `null`). If the `element` can't be found, -1 will be returned.
```

> but `.equals()` method may throw a NullPointerException if the element is null.

This doesn't seem to bear out for this method. After troubleshooting, not only can you add `null` to an `ArrayList` object, the `.indexOf()` will return the index of that `null` value as expect.

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:39]:
```suggestion

```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:47]:
```suggestion

```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:30]:
```suggestion

The following example features two calls to `.indexOf()` on `animals`. The first will print the index of the first occurrence of the element with the string `"Tiger"`. The second will return -1 because `"Elephant"` does not exist in `animals`:

```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:48]:
```suggestion
    System.out.println(animals.indexOf("Tiger"));
```

With the `\n`, the result would be -1.

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:57]:
```suggestion
This will print the following output:
```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:64]: I'm not sure I agree with this point. Isn't it more so that `.indexOf()` is not compatible with _regular arrays_ because it is an `ArrayList` method?

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:66]: Based on my comment directly above this, I'm not sure we even need this second example.

**Review by Dusch4593 — CHANGES_REQUESTED** [review]
@Pavan-Nambi Finished and left you some comments/suggestions 😄

To answer your question about Rust, I'm not entirely sure. I can certainly check in with my colleagues and follow up.

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:3]: ok

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:15]: ya makes sense

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:27]: i will recheck and update

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:64]:
```
// Online Java Compiler
// Use this editor to write, compile and run your Java code online
import java.util.ArrayList;
class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello, World!");
        ArrayList<int> list = new ArrayList<>();  
        list.add(1);
        list.add(2);
        list.add(3);

int index = list.indexOf(2);  
System.out.println(index);

        
    }
}
```
##Output wud be err
```
javac /tmp/CmHIzfwn03/HelloWorld.java
/tmp/CmHIzfwn03/HelloWorld.java:7: error: unexpected type
ArrayList<int> list = new ArrayList<>();  
                  ^
  required: reference
  found:    int
1 error

```

Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:64]:
```
// Online Java Compiler
// Use this editor to write, compile and run your Java code online
import java.util.ArrayList;
class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello, World!");
        ArrayList<Integer> list = new ArrayList<>();  
        list.add(1);
        list.add(2);
        list.add(3);

int index = list.indexOf(2);  
System.out.println(index);

        
    }
}
```
##output normal as we r using integer
```
java -cp /tmp/CmHIzfwn03 HelloWorld
Hello, World!1
```

Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:64]: Integer is an object whereas int is primitive data type
below pic is from official java docs
image
as you can see what .indexOf supports is object not a primitive data type as mentioned in docs...

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:66]: ok ig updating code will do? like to make it show we cant use int in arraylist and use .indexof ?

Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:66]: haa my bad i used array.indexOf tht was a mistake as array dont have that method i misinterpreted docs it should indexof method dont accept primitives

**Comment by Pavan-Nambi** [comment]
hey let me know your opinion after reading comments i will update once finally as it dont make sense to update now and od again after ur comments
thankyou!

**Comment by Pavan-Nambi** [comment]
and it would be nice if you can check in about rust too i am pretty sure there are many ppl who would like to learn/contribute to rust..

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:78]:
```suggestion

```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:23]:
```suggestion
The index of its first occurrence (even if the value if `null`) is returned if the `element` exists. If the `element` can't be found, -1 will be returned.
```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:15]:
```suggestion
The **`.indexOf()`** method returns the index of the first occurrence of the specified element in an `ArrayList`. If the element is not found, -1 is returned.
```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:88]:
```suggestion
```

I don't think it's necessary to explicitly say this is not a regular array method and show the error the will occur.

**Review by Dusch4593 — CHANGES_REQUESTED** [review]
@Pavan-NameA few more comments/suggestions 😄

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:9]: oki

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:72]:
```suggestion
  public static void main(String[] args) {
    int[] arr = {1, 2, 3, 4, 5};

    System.out.println(arr.indexOf(3));
```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:76]: Could we add the following text between lines # 75 and # 76?

Since arrays do not have an `.indexOf()` method, the following error will be thrown:

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Dusch4593) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:85]:
```suggestion
Wrapper classes must be used in order for the `.indexOf()` method to work with primitive values.
```

For Docs entries, let's refrain from using "you", "your", and "you're" and aim for a more neutral, manual-like voice.

Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:85]: ok

**Comment by Pavan-Nambi** [comment]
@Dusch4593 hey there apologies for the delay 
Can you look into that i have made all the requested changes...

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:127]: ya added

Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:127]: @Dusch4593  if we add these lines in there we have to keep  this example 2 one as it is can you look into that? i am confused on whether to remove example 2 or not 

imo its better to remove as now i added another 2 examples on why arraylist.indexof method dont accept int but does integer 

look into that and let me know 

thankyou
https://github.com/Codecademy/docs/pull/1454/files#r1062497759

Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:127]: that link refers to line where you said its better to remove example 2

**Review by Dusch4593 — APPROVED** [review]
@Pavan-Nambi Thank you for implementing that error-prone example into the first "## Example" section! 👍🏻 

I went ahead and removed the second example (02c1ee5e72bbd8b69a2d6c4e807deb7623acd3cb) to cut down on repetition. Thing LGTM for the next review! 😄

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (HishamT) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:29]:
```suggestion
The following example features two calls to `.indexOf()` on an `ArrayList` called `animals`. The first call will print the index of the first occurrence of the element  `"Tiger"`. The second will return -1 because the element `"Elephant"` does not exist:
```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (HishamT) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:38]:
```suggestion
ArrayList<String> animals = new ArrayList<>();
```

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (HishamT) [inline content/java/concepts/array-list/terms/indexOf/indexOf.md:93]:
```suggestion
Wrapper classes (e.g., `Integer`) must be used for the `.indexOf()` method to work with primitive values:
```

**Review by HishamT — CHANGES_REQUESTED** [review]
Hey @Pavan-Nambi, I have finished the second review of your entry and left some comments.

File: content/java/concepts/array-list/terms/indexOf/indexOf.md
Comment (Pavan-Nambi) [reply content/java/concepts/array-list/terms/indexOf/indexOf.md:93]: ok

**Comment by Pavan-Nambi** [comment]
@Dusch4593  @HishamT  can you look into that made requested changes

**Review by HishamT — APPROVED** [review]
LGTM!

**Review by yangc95 — APPROVED** [review]
@Pavan-Nambi This looks good to me!

@HishamT is there anything else that you think may need improving here? I was also thinking about wrapper classes since we don't have anything on them in the Java topic. Would you think it could be a good addition to the data types entry?

**Comment by HishamT** [comment]
Hey @yangc95, yeah, I think we could add the concept of wrapper classes to the Java topic as it is essential for many Java concepts. I will add it to the list of issues I am going to create for Java. :smile:

**Comment by SSwiniarski** [comment]
@Pavan-Nambi congrats on your new entry! You can see it live here: https://www.codecademy.com/resources/docs/java/array-list/indexOf
image

**Pull request merged**
